### ISSUE TITLE
Empty Vec causes SIGSEGV during deserialization

### ISSUE THREAD

**Issue by mahou-shoujo** [djkoloski/rkyv#89]
Hello, the following example produces a SIGSEGV while deserializing the archived data (rkyv 0.4.2, default features).
```
#[test]
fn boom() {
    use rkyv::{
        archived_value,
        de::deserializers::AllocDeserializer,
        ser::{serializers::WriteSerializer, Serializer},
        Archive, Deserialize, Serialize,
    };
    #[derive(Archive, Deserialize, Serialize, Debug, PartialEq)]
    struct Test {
        vec: Vec<i32>,
    }
    let value = Test { vec: vec![] };

    let mut serializer = WriteSerializer::new(Vec::new());
    let pos = serializer
        .serialize_value(&value)
        .expect("failed to serialize value");
    let buf = serializer.into_inner();

    let archived = unsafe { archived_value::<Test>(buf.as_ref(), pos) };

    let mut deserializer = AllocDeserializer;
    let deserialized = archived
        .deserialize(&mut deserializer)
        .expect("failed to deserialize value");
    assert_eq!(deserialized, value);
}
```
A similar test case does work with version 0.3.1 though and it seems like the regression happened in 0.4.0. My rustc profile:
```
debug_assertions
target_arch="x86_64"
target_endian="little"
target_env="gnu"
target_family="unix"
target_feature="fxsr"
target_feature="sse"
target_feature="sse2"
target_os="linux"
target_pointer_width="64"
target_vendor="unknown"
unix
```

**Comment by djkoloski** [comment]
Can you run your buffer through [`bytecheck`](https://github.com/djkoloski/bytecheck) and see if you get an error?

Example:

```rust
// ...

let archived = rkyv::validation::check_archive::<Test>(buf.as_ref(), pos).expect("check bytes failed");

// ...
```

This could be an alignment issue like #83. The docs need a pass to emphasize that buffers need to be properly aligned.

**Comment by mahou-shoujo** [comment]
Yes, `check_archive` also fails with the same memory error (even though it is not an unsafe function). I would think Vec buffers will always have some good alignment on my target but just to get it out of the way I had tried to use `Aligned<[u8; 256]>` as buffer (+ `BufferSerializer`) and still got the same error. 

You can find the code in [this repo](https://github.com/mahou-shoujo/rkyv-test-case-rs). When I made it, I realized the test only fails if I use [jemallocator](https://crates.io/crates/jemallocator) and works just fine with the default one. Perhaps jemalloc is more pedantic?

**Comment by djkoloski** [comment]
Thank you for the detailed repro, I'll take a look at this as soon as I can. I can't think of anything that would cause it to fail by changing the default allocator to jemalloc off the top of my head.

**Comment by djkoloski** [comment]
Finally got a chance to track this down. This was due to a pretty subtle issue, so thank you very much for your help.

In the documentation for [`GlobalAlloc::alloc`](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#tymethod.alloc):

> ## Safety
>
> This function is unsafe because undefined behavior can result if the caller does not ensure that layout has non-zero size.

The `Deserialize` implementation for slices was not respecting this requirement, and requested allocations with layouts of zero size when the slice had length 0. This has been fixed in c803c1b288472284f09611b72914ba5c794206e9 by checking if the length of the slice is 0 and returning a null pointer.

The fix is available as of [0.4.3](https://github.com/djkoloski/rkyv/releases/tag/v0.4.3).
